aiographite/aiographitesend.py

```diff
@@ -0,0 +1,449 @@
+		else:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value, timestamp) for metric_dir_list, value, timestamp in dataset]
+
+		self.send_valid_dataset_list(valid_dataset, timestamp)
+
+
+
+	def send_single_valid_data(self, metric, value, timestamp = None):
+		"""
+			@metric: String
+			@value: int
+			@timestamp: int
+			Send a single data(metric value timestamp) to graphite
+		"""
+		timestamp = int(time.time()) if timestamp is None else int(timestamp)
+		message = ""
+		if self.protocol == "plaintext":
+			message = plaintext_protocol_formatted_data(metric, value, timestamp)
+		else:
+			listOfMetricTuples = [pickle_protocol_formatted_data(metric, value, timestamp)]
+			message = generate_message_for_pickle(listOfMetricTuples)
+		self.loop.run_until_complete(asyncio.ensure_future(self.send_message(message)))
+
+
+
+	def send_valid_dataset_list(self, dataset, timestamp = None):
+		"""
+			@param: 
+			Support two kinds of dataset
+				1)	dataset = [(metric1, value1), (metric2, value2), ...] 
+				or 
+				2)	dataset = [(metric1, value1, timestamp1), (metric2, value2, timestamp2), ...]
+		"""
+		timestamp = int(time.time()) if timestamp is None else int(timestamp)
+		message = ""
+
+		# Generate message based on protocol
+		# 1. plaintext
+		# 2. pickle
+		if self.protocol == "plaintext":
+			message = self.generate_message_for_data_list(dataset, timestamp, self.plaintext_protocol_formatted_data, self.generate_message_for_plaintext)
+		else:
+			message = self.generate_message_for_data_list(dataset, timestamp, self.pickle_protocol_formatted_data, self.generate_message_for_pickle)
+
+		# Sending Data
+		self.loop.run_until_complete(asyncio.ensure_future(self.send_message(message)))
+
+
+
+	def send_valid_dataset_dict(self, dataset, timestamp = None):
+		"""
+			Send data to graphite server when incoming data is in 'dict' format
+			@param: dataset = {
+									metric1 : value1,      // type ( string: int )
+									metric2 : value2, 
+									...
+							  }
+
+			metric1 (metric2, ...) are valid metric name for Graphite
+		"""
+		self.send_valid_dataset_list(dataset.items(), timestamp)
+
+
+
+	def generate_message_for_data_list(self, dataset, timestamp, formate_function, generate_message_function):
+		"""
+			generate proper formatted message 
+			@param:
+			Support two kinds of dataset
+				1)	dataset = [(metric1, value1), (metric2, value2), ...] 
+				or 
+				2)	dataset = [(metric1, value1, timestamp1), (metric2, value2, timestamp2), ...]
+		"""
+		listofData = []
+		for data in dataset:
+			# unpack metric data
+			if len(data) == 2:
+				(metric, value) = data
+			else:
+				(metric, value, data_timestamp) = data
+				timestamp = data_timestamp
+			listOfData.append(formate_function(metric, value, timestamp))
+		message =  generate_message_function(listofData)
+		return message
+
+
+
+	def generate_message_for_plaintext(self, listOfPlaintext):
```

Comment:
I see that for plaintext and pickle, there's some custom commands we have to run for those:

```
generate_message
formatted_data
```
If I see a lot of behavior that depends on an enum value, we can convert those to classes, and pass in classes instead:

```
graphite_plaintext = AIOGraphite(protocol=PlaintextProtocol())
graphite_pickle = AIOGraphite(protocol=PickleProtocol())
```

this is also very cool, because it allows other developers to pass in their own pickle format as well.